src/GafferScene/MeshTangents.cpp

```diff
@@ -0,0 +1,155 @@
+	addChild( new StringPlug( "uvSet", Plug::In, "st" ) );
+	addChild( new StringPlug( "posPrimVarName", Plug::In, "P" ) );
+	addChild( new BoolPlug( "orthogonal", Plug::In, true ) );
+
+	// Fast pass-throughs for things we don't modify
+	outPlug()->attributesPlug()->setInput( inPlug()->attributesPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+}
+
+MeshTangents::~MeshTangents()
+{
+}
+
+Gaffer::StringPlug *MeshTangents::uvSetPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *MeshTangents::uvSetPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug *MeshTangents::posPrimVarNamePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *MeshTangents::posPrimVarNamePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::BoolPlug *MeshTangents::orthogonalPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+const Gaffer::BoolPlug *MeshTangents::orthogonalPlug() const
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+void MeshTangents::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneElementProcessor::affects( input, outputs );
+
+	if( input == uvSetPlug() || input == posPrimVarNamePlug() || input == orthogonalPlug() )
+	{
+		outputs.push_back( outPlug()->objectPlug() );
+	}
+	else if( input == outPlug()->objectPlug() )
+	{
+		outputs.push_back( outPlug()->boundPlug() );
+	}
+}
+
+bool MeshTangents::processesBound() const
+{
+	return true;
+}
+
+void MeshTangents::hashProcessedBound( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	hashProcessedObject( path, context, h );
+}
+
+Imath::Box3f MeshTangents::computeProcessedBound( const ScenePath &path, const Gaffer::Context *context, const Imath::Box3f &inputBound ) const
+{
+	return inputBound;
+}
+
+bool MeshTangents::processesObject() const
+{
+	return true;
+}
+
+void MeshTangents::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	uvSetPlug()->hash( h );
+	posPrimVarNamePlug()->hash( h );
+	orthogonalPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr MeshTangents::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::ConstObjectPtr inputObject ) const
+{
+	const MeshPrimitive *mesh = runTimeCast<const MeshPrimitive>( inputObject.get() );
+	if( !mesh )
+	{
+		return inputObject;
+	}
+
+	std::string uvSet = uvSetPlug()->getValue();
+	std::string posPrimVarName = posPrimVarNamePlug()->getValue();
+	bool ortho = orthogonalPlug()->getValue();
+
+	std::pair<PrimitiveVariable, PrimitiveVariable> tangentPrimvars = MeshAlgo::calculateTangents( mesh, uvSet, ortho, posPrimVarName);
+	MeshPrimitivePtr meshWithTangents = runTimeCast<MeshPrimitive>( mesh->copy() );
+
+	meshWithTangents->variables["Tangent"] = tangentPrimvars.first;
```

Comment:
> I suppose I'm using the PR as mechanism to propose speculative code changes and follow the discussion, perhaps this is an abuse of the feature and frustrating.

I think it's a pretty decent way of discussing things given we can't do it face to face. It'd just be good to know what's speculative and what's not - maybe a second commit with the speculative bit would have helped isolate it for discussion, and make it easy to roll back or squash it in later.

> What does the phrase "along for the ride" mean in regards to primitive variables?

Sorry, I didn't explain myself very well. I was thinking of two things :

- Once created, a primvar stays as it is - it's just data being passed along. If the mesh was deformed downstream, nothing would magically update the tangents.
- We don't currently assign any meaning to the names of primvars, other than in a couple of spots in the renderer backends (we map s/t to Arnold's uv attribute for instance). Generally (I think I'm right in saying this, but perhaps I should take a look at some of the internal nodes) nodes get told what primvar to operate on, rather than looking for predefined names.

> In retrospect what you describe sounds simple the for the user to understand, general & cleaner API. I suppose I just went on tangent with regards to the derivative notation.

I _think_ it is. I hesitated a bit while writing the above, because in some ways it'd be nice for the downstream nodes to rely on some naming semantics, so they can be like "find the tangents primvars, or compute them if they're not there". But I think that falls down because of the lack of automatic updates when things are deformed - if primvars are just along for the ride, they can't really have well defined semantics. So I think exposing the names to the user on plugs on both the producing and consuming nodes is the best we can do, and as long the default values line up for those, folks won't need to touch them most of the time.

> Should I wait for further discussion from other parties before implementing the following:

>    add uTangentPlug() & vTangentPlug() which default to uTangent & vTangent. I assume were going to be consistent using uv rather than st in cortex 10?

If nobody want to champion automatic semantic naming, then I still think that would be my preference.

If we do have plugs, then I definitely prefer those plug names over the other alternatives - they're reasonably non-technical, and make it clear that the tangents are with respect to the uv set. I'm not 100% sure whether st should be renamed for Cortex 10, but those default values sound fine to me too. If you do make this change, could you update the docs so that we consistently use uTangent/vTangent terminology rather than mix and match with the tangent/bitangent terms?